### DISCUSSION TITLE
Compile error after changing `examples/uretprobe` 's target from amd64 to arm64.

### DISCUSSION THREAD

**Discussion by xiaoziv** [cilium/ebpf#1122] [Q&A] [answered]
In file `examples/uretprobe/main.go`, I changed the code `//go:build amd64` to `//go:build arm64` in line 5, and change the target parameter from `-target amd64` to `-target arm64`.

Error occurs when I rerun the `go generate` command, here is the error:

```shell
uretprobe git:(main) ✗ go generate
/Volumes/TuyaExt/work/github/ebpf/examples/uretprobe/uretprobe.c:26:58: error: incomplete definition of type 'struct user_pt_regs'
        bpf_probe_read(&event.line, sizeof(event.line), (void *)PT_REGS_RC(ctx));
                                                                ^~~~~~~~~~~~~~~
../headers/bpf_tracing.h:209:46: note: expanded from macro 'PT_REGS_RC'
#define PT_REGS_RC(x) (((PT_REGS_ARM64 *)(x))->regs[0])
                       ~~~~~~~~~~~~~~~~~~~~~~^
/Volumes/TuyaExt/work/github/ebpf/examples/uretprobe/uretprobe.c:26:58: note: forward declaration of 'struct user_pt_regs'
../headers/bpf_tracing.h:209:26: note: expanded from macro 'PT_REGS_RC'
#define PT_REGS_RC(x) (((PT_REGS_ARM64 *)(x))->regs[0])
                         ^
../headers/bpf_tracing.h:200:45: note: expanded from macro 'PT_REGS_ARM64'
#define PT_REGS_ARM64 const volatile struct user_pt_regs
                                            ^
1 error generated.
Error: can't execute clang: exit status 1
exit status 1
main.go:25: running "go": exit status 1
```

Does something wrong with the definition of `PT_REGS_RC` in arm64 architecture in headers/bpf_tracing.h file?

**Comment by florianl** [comment]
In [`examples/headers/bpf_tracing.h`](https://github.com/cilium/ebpf/blob/01ebd4c1e2b9f8b3dd4fd2382aa1092c3c9bfc9d/examples/headers/bpf_tracing.h#L198-L200) you see, that `arm64` uses `user_pt_regs` instead of `pt_regs`. 

https://github.com/cilium/ebpf/blob/01ebd4c1e2b9f8b3dd4fd2382aa1092c3c9bfc9d/examples/headers/bpf_tracing.h#L198-L200

But in [`examples/headers/common.h`](https://github.com/cilium/ebpf/blob/01ebd4c1e2b9f8b3dd4fd2382aa1092c3c9bfc9d/examples/headers/common.h#L113-L114) only `struct pt_regs` for `aarch64` is defined.
https://github.com/cilium/ebpf/blob/01ebd4c1e2b9f8b3dd4fd2382aa1092c3c9bfc9d/examples/headers/common.h#L113-L114
 The definition for `struct user_pt_regs` on `arm64` is not provided in the headers. Overall, I think, the examples do have a focus on `aarch64`.

**Reply by xiaoziv** [reply]
Sorry for my unfamiliarity with the c macro code.
I want to use `PT_REGS_RC` to extract return value of kretprobe, the code may like

```c
SEC("kretprobe/tcp_sendpage")
int kretprobe__tcp_sendpage(struct pt_regs *ctx)
{
    int size = PT_REGS_RC(ctx);
    if (size > 0)
        return tcp_sendstat(size);
    else
        return 0;
}
```

Furthermore, I want to compile the code into both x86 and arm64 architecture. 
Is there any way to approach this as `PT_REGS_RC` doesn't work well in arm64?

**Comment by xiaoziv** [comment, marked as answer]
I have solved this problem by appending the following snippet to `headers/common.h`.
@florianl thanks.

```
#if defined(__TARGET_ARCH_arm64)
struct user_pt_regs {
	__u64		regs[31];
	__u64		sp;
	__u64		pc;
	__u64		pstate;
};
struct pt_regs {
	union {
		struct user_pt_regs user_regs;
		struct {
			u64 regs[31];
			u64 sp;
			u64 pc;
			u64 pstate;
		};
	};
	u64 orig_x0;
#ifdef __AARCH64EB__
	u32 unused2;
	s32 syscallno;
#else
	s32 syscallno;
	u32 unused2;
#endif
	u64 sdei_ttbr1;
	/* Only valid when ARM64_HAS_GIC_PRIO_MASKING is enabled. */
	u64 pmr_save;
	u64 stackframe[2];

	/* Only valid for some EL1 exceptions. */
	u64 lockdep_hardirqs;
	u64 exit_rcu;
};
#endif
```
